packages/axes/src/compute.ts

```diff
@@ -197,43 +211,60 @@ export const computeCartesianTicks = ({
     }
 }
 
-export const getFormatter = (format, scale) => {
-    if (!format || typeof format === 'function') return format
+export const getFormatter = <Value extends string | number | Date>(
+    format: string | ValueFormatter<Value> | undefined,
+    scale: AnyScale
+): ValueFormatter<Value> | undefined => {
+    if (typeof format === 'undefined' || typeof format === 'function') return format
 
     if (scale.type === 'time') {
-        const f = timeFormat(format)
-        return d => f(new Date(d))
+        const formatter = timeFormat(format)
+
+        return (d => formatter(d instanceof Date ? d : new Date(d))) as ValueFormatter<Value>
     }
 
-    return d3Format(format)
+    return (d3Format(format) as unknown) as ValueFormatter<Value>
 }
 
-export const computeGridLines = ({ width, height, scale, axis, values: _values }) => {
-    const lineValues = Array.isArray(_values) ? _values : undefined
+export const computeGridLines = <Value extends string | number | Date>({
+    width,
+    height,
+    scale,
+    axis,
+    values: _values,
+}: {
+    width: number
+    height: number
+    scale: AnyScale
+    axis: 'x' | 'y'
+    values?: TicksSpec<Value>
+}) => {
+    const lineValues = isArray<number>(_values) ? _values : undefined
     const lineCount = isInteger(_values) ? _values : undefined
 
     const values = lineValues || getScaleTicks(scale, lineCount)
 
-    const position = scale.bandwidth ? centerScale(scale) : scale
-
-    let lines
-    if (axis === 'x') {
-        lines = values.map(v => ({
-            key: `${v}`,
-            x1: position(v),
-            x2: position(v),
-            y1: 0,
-            y2: height,
-        }))
-    } else if (axis === 'y') {
-        lines = values.map(v => ({
-            key: `${v}`,
-            x1: 0,
-            x2: width,
-            y1: position(v),
-            y2: position(v),
-        }))
-    }
+    const position = 'bandwidth' in scale ? centerScale(scale) : scale
+
+    const lines: Line[] = values.map(value => {
```

Comment:
nit: the `if` outside was to avoid running this check for each item, as it's stable.